### ISSUE TITLE
Option to use AMBEserver

### ISSUE THREAD

**Issue by csylvain** [f4exb/sdrangel#266]
feature/enhancement request to offer to use a local AMBEserver and avoid all copyright/license issues when decoding D-Star/Fusion/DMR(ETSI 102 361)/other transmissions. Recommend use of products containing an AMBE-3000 chip supported by AMBEserver.

**Comment by f4exb** [comment]
AMBE 3000 is already supported (and has been for a long time) as an USB dongle like the thumbDV using the SerialDV library: https://github.com/f4exb/serialDV See the compile Wiki to know how to compile with SerialDV support. It is also mentioned on the main readme that would be a good idea for you to read.

So no need for this.

**Comment by csylvain** [comment]
well, i swear i did read through all of the documentation (even the developer docs) at https://github.com/f4exb/sdrangel/ and since i didn't encounter any mention of ThumbDV (or breadcrumbs leading me to serialDV to discover more features), i'm sorry i did not realize this.
however, the advantage to using AMBEserver is one ThumbDV (or similar) can serve all the decoding needs within a ham's shack (within the dongle's bitrate limitation, of course).
as is now, the dongle must be dedicated to only the one SDRangel instance on one system.

**Comment by f4exb** [comment]
In this paragraph: https://github.com/f4exb/sdrangel#dsd-digital-speech-decoder
"If you have one or more serial devices interfacing the AMBE3000 chip in packet mode you can use them to decode AMBE voice frames. For that purpose you will need to compile with SerialDV support..."

So what you suggest is to add support for "your" AMBEserver which sources are difficult to dig out of the rest and the protocol only to be found by deciphering the source code. I am not sure I would like to do that and at least not with high priority.

**Comment by csylvain** [comment]
thank you. after your first suggestion in this thread, i did a text search of the SDRAngel readme and found the mention in the midst of a middle paragraph of the DSD plugin section, surrounded by the discussion of building the plugin and the legalities in doing so. it may well be advantageous to rewrite the section to make it clear what the plugin itself is or is not and also to separate the support of an AMBE chip into its own featured section.

agreed the prospect of digging through the DL5DI OpenDV repo to make sense of the relevant C++ code does not appear to be a pleasant prospect.

however, a much simpler example of working AMBEserver code can be found at https://github.com/marrold/AMBEServer .. perhaps this would provide the best guidance?

there is, as you probably well know, an existing AMBEServer ecology for D-Star/Fusion/DMR and its advantages are the motivation for this enhancement request. Additionally, offering the capability would provide a technical solution for the 'AMBE problem' and protective legal legerdemain for everyone.

**Comment by f4exb** [comment]
Where else could it be outside the paragraph about DSD? I don't  know. 

And yes there are legality issues around AMBE which are worth to mention apart from the fact that the AMBE dongle will give you results largely better than the soft MBElib for audio quality and also because I think MBElib ignores the embedded FEC.

You can already use an AMBE dongle (or any number of them) with SDRangel. The main point of the server is to make the dongle(s) accessible to more than one application which is indeed interesting but not vital hence my statement on the low priority of such a change because the ROI is not so great.

**Comment by n8ohu** [comment]
Having the ability to use AMBEserver and a PiDV (the version of the Northwest Digital Radio ThumbDV that mounted on the GPIO pins of the Raspberry Pi) would be useful for those of us that have them.

**Comment by f4exb** [comment]
I am not too keen on doing something special for the Pi. It would be useful only if SDRangel (hence the server variant) is run on the Pi.

**Comment by csylvain** [comment]
that is certainly true for the PiDV. AMBEserver is not narrowly applicable,
however.

**Comment by f4exb** [comment]
From the source code of the server it seems the packet that would be sent over the serial link to the AMBE3000 is just sent to the UDP port on which the AMBEserver listens. Below the relevant extracts of the code:
```c
static const unsigned char DV3K_START_BYTE   = 0x61;
#define dv3k_packet_size(a) (1 + sizeof((a).header) + ntohs((a).header.payload_length))

#pragma pack(push, 1)
struct dv3k_packet {
    unsigned char start_byte;
    struct {
        unsigned short payload_length;
        unsigned char packet_type;
    } header;
    union {
        struct {
            unsigned char field_id;
            union {
                char prodid[16];
                char ratep[12];
                char version[48];
                short chanfmt;
            } data;
        } ctrl;
        struct {
            unsigned char field_id;
            unsigned char num_samples;
            short samples[160];
            unsigned char cmode_field_id;
            short cmode_value;
        } audio;
        struct {
            struct {
                unsigned char field_id;
                unsigned char num_bits;
                unsigned char data[9];
            } data;
            struct {
                unsigned char field_id;
                unsigned short value;
            } cmode;
            struct {
                unsigned char field_id;
                unsigned char tone;
                unsigned char amplitude;
            } tone;
        } ambe;
    } payload;
};
#pragma pack(pop)


int processSocket(int sockFd, int serialFd)
{
	socklen_t addrLen = sizeof(struct sockaddr_in);
	ssize_t bytesRead;
	struct dv3k_packet packet;
	
	bytesRead = recvfrom(sockFd, &packet, sizeof(packet), 0, (struct sockaddr *) &sa, &addrLen);
	if (bytesRead < 0) {
		fprintf(stderr, "AMBEserver: error when reading from the socket: %s\n", strerror(errno));
		return 0;
	}

	if (debug)
		dump("Socket data", &packet, bytesRead);

	if (packet.start_byte != DV3K_START_BYTE) {
		fprintf(stderr, "AMBEserver: invalid start byte when reading from the socket, 0x%02X", packet.start_byte);
		return 1;
	}

	if (bytesRead != dv3k_packet_size(packet)) {
		fprintf(stderr, "AMBEserver: invalid length when reading from the socket, %zd=%zd", bytesRead, dv3k_packet_size(packet));
		return 1;
	}

	if (write(serialFd, &packet, dv3k_packet_size(packet)) != dv3k_packet_size(packet)) {
		fprintf(stderr, "AMBEserver: error when writing to the serial port: %s\n", strerror(errno));
		return 0;
	}

	return 1;
}
```

**Comment by f4exb** [comment]
Then it looks like the interface with the UDP server should be done in serialdv: https://github.com/f4exb/serialDV extending its capabilities to UDP with the option to open the UDP link instead of the serial link. Then this opening method will be used in `dvserialworker.cpp` when UDP is requested instead of a direct serial link.

**Comment by csylvain** [comment]
i saw the code was direct to udp with very little (none, really) 'bit
handling'. having extended SDRAngel with this, then one ThumbDV with
AMBEserver can turn datastream into audio for more than one instance of
SDRAngel and/or for other software on the same computer or LAN (up to the
maximum bandwidth of USB or the AMBE-3000 chip).

 - 73 -

**Comment by f4exb** [comment]
Implemented in v4.11.2

**Comment by csylvain** [comment]
awesome! thanks!
